price-feeder/telemetry/metrics.go

```diff
@@ -0,0 +1,173 @@
+	"github.com/prometheus/common/expfmt"
+)
+
+// globalLabels defines the set of global labels that will be applied to all
+// metrics emitted using the telemetry package function wrappers.
+var globalLabels = []metrics.Label{}
+
+// Metrics supported format types.
+const (
+	FormatDefault    = ""
+	FormatPrometheus = "prometheus"
+	FormatText       = "text"
+)
+
+// Config defines the configuration options for application telemetry.
+type Config struct {
+	// Prefixed with keys to separate services
+	ServiceName string `toml:"service_name"`
+
+	// Enabled enables the application telemetry functionality. When enabled,
+	// an in-memory sink is also enabled by default. Operators may also enabled
+	// other sinks such as Prometheus.
+	Enabled bool `toml:"enabled"`
+
+	// Enable prefixing gauge values with hostname
+	EnableHostname bool `toml:"enable_hostname"`
+
+	// Enable adding hostname to labels
+	EnableHostnameLabel bool `toml:"enable_hostname_label"`
+
+	// Enable adding service to labels
+	EnableServiceLabel bool `toml:"enable_service_label"`
+
+	// PrometheusRetentionTime, when positive, enables a Prometheus metrics sink.
+	// It defines the retention duration in seconds.
+	PrometheusRetentionTime int64 `toml:"prometheus_retention_time"`
+
+	// GlobalLabels defines a global set of name/value label tuples applied to all
+	// metrics emitted using the wrapper functions defined in telemetry package.
+	//
+	// Example:
+	// [["chain_id", "cosmoshub-1"]]
+	GlobalLabels [][]string `toml:"global_labels"`
+}
+
+// Metrics defines a wrapper around application telemetry functionality. It allows
+// metrics to be gathered at any point in time. When creating a Metrics object,
+// internally, a global metrics is registered with a set of sinks as configured
+// by the operator. In addition to the sinks, when a process gets a SIGUSR1, a
+// dump of formatted recent metrics will be sent to STDERR.
+type Metrics struct {
+	memSink           *metrics.InmemSink
+	prometheusEnabled bool
+}
+
+// GatherResponse is the response type of registered metrics
+type GatherResponse struct {
+	Metrics     []byte
+	ContentType string
+}
+
+// New creates a new instance of Metrics
+func New(cfg Config) (*Metrics, error) {
+	if !cfg.Enabled {
+		return nil, nil
+	}
+
+	if numGlobalLables := len(cfg.GlobalLabels); numGlobalLables > 0 {
+		parsedGlobalLabels := make([]metrics.Label, numGlobalLables)
+		for i, gl := range cfg.GlobalLabels {
+			parsedGlobalLabels[i] = NewLabel(gl[0], gl[1])
+		}
+
+		globalLabels = parsedGlobalLabels
+	}
+
+	metricsConf := metrics.DefaultConfig(cfg.ServiceName)
+	metricsConf.EnableHostname = cfg.EnableHostname
+	metricsConf.EnableHostnameLabel = cfg.EnableHostnameLabel
+
+	memSink := metrics.NewInmemSink(10*time.Second, time.Minute)
+	metrics.DefaultInmemSignal(memSink)
+
+	m := &Metrics{memSink: memSink}
+	fanout := metrics.FanoutSink{memSink}
+
+	if cfg.PrometheusRetentionTime > 0 {
+		m.prometheusEnabled = true
```

Comment:
So I know this was ported from the SDK (which I wrote actually), but realizing all this time later, I think there is a slightly cleaner way for operators. I would suggest we do the following:

1. Remove the `PrometheusRetentionTime` config.
2. Add a `Type` config instead, whose value must be `generic` or `prometheus`.
3. Hardcode `PrometheusRetentionTime` to something like 120 or whatever.